### PR TITLE
Simplify html-block and MathJax rendering to use Lit and avoid temp element creation

### PR DESCRIPTION
**Pull request by grant-cleary**
The goal with this PR is to try to simplify some of the core html-block rendering logic (and the MathJax renderer logic) to align more with our standard Lit components and avoid creating chains of temp elements that could cause trouble for us.

To do this, we use Lit to render the barebones render container and the slot we use for stamping, but we still do all of the rendering itself manually outside of the Lit rendering lifecycle. This is required because of how MathJax in particular needs the elements it's working on to be attached to the DOM when it does its work. i.e. We need the html-block contents to be stamped into the shadowRoot before we can hand off to MathJax for rendering.

### CODE REVIEW

File: components/html-block/html-block.js
Comment (grant-cleary) [inline components/html-block/html-block.js:241]: There's some jank here because (again) things get really complicated if we need to dynamically change MathJax's configuration. So this logic works in a general sense, but once the MathJax context object is set to something valid the first time, subsequent changes won't accomplish anything.

This isn't ideal, but it's not a change in behaviour and I don't want to muddy this PR with additional work to support dynamically changing MathJax's configuration.

File: components/html-block/html-block.js
Comment (grant-cleary) [inline components/html-block/html-block.js:299]: Wipe the render container before we start stamping. We can't assume that whatever was rendered here before stamping is still valid. Prism and MathJax both effectively need us to start from a clean slate, for example.

File: components/html-block/html-block.js
Comment (grant-cleary) [inline components/html-block/html-block.js:303]: It's important that we stamp the slot contents into the render container **before** we run any of the renderers. MathJax in particular requires that these be attached to the DOM or they won't typeset properly.

File: helpers/mathjax.js
Comment (grant-cleary) [inline helpers/mathjax.js:120]: This handles the Lit comment problem. The reason that exploded in our custom adapter here is because the adaptor is expecting a structure that looks exactly like this:
```
<div class="d2l-html-block-rendered">
   <mjx-doc>
      <mjx-head>...</mjx-head>
      <mjx-body>...</mjx-body>
   </mjx-doc>
</div>
```
Effectively, we use these `mjx-{structure}` nodes to simulate an HTML document for MathJax to process. Lit messes things up here because it produces a structure like this (note the comment):
```
<div class="d2l-html-block-rendered">
   <!--lit$1425314-->
   <mjx-doc>...</mjx-doc>
</div>
```
So the solution implemented here is to just deliberately ignore non-element nodes when our adaptor parses the document structure.

File: helpers/mathjax.js
Comment (grant-cleary) [inline helpers/mathjax.js:70]: Unfortunately we can't really avoid tampering with the element's contents again here because we do need to build this structure so our custom adaptor has a "document" to hook onto.

File: helpers/mathjax.js
Comment (grant-cleary) [inline helpers/mathjax.js:125]: This implementation does keep our custom adaptor. Brief forays into removing it have so far proven fruitless, so I'm not optimistic we can ditch it. MathJax just stores too much document-level state. However, I will keep poking at it to see if there's a path forward where we can just move to their default APIs.

File: components/html-block/test/html-block.test.js
Comment (grant-cleary) [inline components/html-block/test/html-block.test.js:55]: This wasn't failing any tests because it was the last renderer in the chain, but not returning the element in this case definitely breaks things now that I've added the extra renderer. 😅

File: helpers/mathjax.js
Comment (grant-cleary) [reply helpers/mathjax.js:120]: Actually, this isn't true now that I'm wiping out the container's inner HTML before stamping. Let me see if I can remove this.

File: components/html-block/html-block.js
Comment (dbatiste) [inline components/html-block/html-block.js:166]: Are these selectors overqualified? Could they just be:

```css
.d2l-html-block-inline {
   ...
}
.d2l-html-block-no-deferred-rendering {
   ...
}
.d2l-html-block-deferred-rendering {
   ...
}
```
Admittedly though, the `slot` element selector provides some hint/clarity.

Comment (grant-cleary) [reply components/html-block/html-block.js:166]: Yeah, they could be simplified. I always struggle at least a bit with how specific to get to avoid potential issues or pitfalls later. I'm content with whatever you would prefer.

**Comment by dbatiste** [comment]
Just an FYI regarding `_renderInline` - it is now safe to use [assignedElements](https://developer.mozilla.org/en-US/docs/Web/API/HTMLSlotElement/assignedElements). You'd still need to find the `DIV`, but filtering on `Node.ELEMENT_NODE` would no longer be necessary if using the element-specific API.

**Comment by grant-cleary** [comment]
> Just an FYI regarding `_renderInline` - it is now safe to use [assignedElements](https://developer.mozilla.org/en-US/docs/Web/API/HTMLSlotElement/assignedElements). You'd still need to find the `DIV`, but filtering on `Node.ELEMENT_NODE` would no longer be necessary if using the element-specific API.

Hmm, is that true? A few weeks ago I was told we still didn't want to break Safari 11.

File: components/html-block/html-block.js
Comment (dbatiste) [reply components/html-block/html-block.js:166]: I'm good either way re the use of the `:host([no-deferred-rendering])` - there is a pro each way imo.

If keeping with the way it is, I'd reduce it to the minimum specificity that's required to avoid unnecessary selector checks (since `element.class-name` is in theory, two selector checks, ignoring possible optimizations the browser may or may not be doing). 🤷‍♂️

File: components/html-block/html-block.js
Comment (dbatiste) [reply components/html-block/html-block.js:241]: Sounds good.  Out of curiosity, do we envision there ever being a use case for it?

File: components/html-block/html-block.js
Comment (grant-cleary) [reply components/html-block/html-block.js:166]: Seems good. The slot has the opposite condition, which I could either accomplish with `:host(:not([no-deferred-rendering])) slot` or by keeping it as is (or I guess by using two separate selectors). Do you have a particular preference there? I think using the generated conditional class name means one selector and otherwise we need two, but... 🤷

File: components/html-block/html-block.js
Comment (grant-cleary) [reply components/html-block/html-block.js:241]: I have a hard time thinking of one. The big reason we have to handle this via a controller is because in FRAs in particular, it's possible for the component to load before ifrau syncs down the attributes from the host page. However, because the MathJax renderer doesn't instantiate MathJax at all if there's no context object present, that case is covered for us and MathJax will init properly when it's first populated.

So... The use case I think we actually care about now isn't impacted by this long-term "caching" of sorts and I'm not sure there's a use case for adjusting the config on the fly.

File: components/html-block/html-block.js
Comment (dbatiste) [reply components/html-block/html-block.js:303]: Almost worth include a comment to that effect... `// must attach nodes before processing renderer plugins since some require connected nodes (ex. MathJax)"`.

File: components/html-block/html-block.js
Comment (grant-cleary) [reply components/html-block/html-block.js:241]: I have a hard time thinking of one. The big reason we have to handle this via a controller is because in FRAs in particular, it's possible for the component to load before ifrau syncs down the attributes from the host page. However, because the MathJax renderer doesn't instantiate MathJax at all if there's no context object present, that case is covered for us and MathJax will init properly when it's first populated.

So... The use case I think we actually care about now isn't impacted by this long-term "caching" of sorts and I'm not sure there's a use case for adjusting the config on the fly.

File: components/html-block/html-block.js
Comment (grant-cleary) [reply components/html-block/html-block.js:303]: Good call, yeah, otherwise this requirement isn't at all obvious and someone could easily come along and switch the order without realizing it's going to subtly break stuff.

File: components/html-block/test/html-block.test.js
Comment (dbatiste) [inline components/html-block/test/html-block.test.js:164]: 🎉 ... I think this is a good additional test.

What are your thoughts on a unit or visual-diff test that includes both the code formatting and the math rendering? Such a test would at least ensure they are properly returning the element, which is currently a requirement of `d2l-html-block` unless we change the API such that plugin renderers are always required to operate on the same element, and not create a new one... not sure if that is possible though. Of course, renderers that are external to core (like @mentions) could still blow things up and would not get caught here, as previously mentioned elsewhere.

File: helpers/mathjax.js
Comment (dbatiste) [inline helpers/mathjax.js:65]: I know this is what we were already doing, but is there a chance that we're blowing away properties that already exist on `mspace`? If so, is it intentional, or should we be using `elm.style.display = 'block';` etc.?

File: helpers/mathjax.js
Comment (dbatiste) [inline helpers/mathjax.js:76]: If I follow this correctly, this now returns the same element that was passed in. Is that correct?

File: helpers/mathjax.js
Comment (dbatiste) [reply helpers/mathjax.js:120]: I think the comment node is what was blowing up when I first tried to use Lit rendering with MathJax in the very beginning.  Glad we found a solution.  :)

File: components/html-block/html-block.js
Comment (dbatiste) [reply components/html-block/html-block.js:166]: I don't feel strongly either way. I'm mostly focused on reducing the selectors to what's necessary, whichever way it's written. I do like relying on single source of truth (the host attribute) and avoiding conditional rendering which adds very slight complexity at the expense of perhaps an extra bit of CSS. I think the `:not()` could be avoided, but I really think it's splitting hairs we can't even measure.

File: helpers/mathjax.js
Comment (grant-cleary) [reply helpers/mathjax.js:120]: It feels a bit fragile, in the sense that if Lit ever decides to put comments elsewhere this might explode... But one of the things that appealed to me about MathJax 3.2 is that they added proper handling for comments _within_ the document body they're parsing. So I think this special handling for direct children of the shadowRoot is all we need.

File: components/html-block/html-block.js
Comment (grant-cleary) [reply components/html-block/html-block.js:166]: Yeah, fair enough.

File: helpers/mathjax.js
Comment (grant-cleary) [reply helpers/mathjax.js:65]: `mspace` shouldn't have any additional properties according to spec, but I think your suggestion is safer regardless so I'll just go with that.

File: helpers/mathjax.js
Comment (grant-cleary) [reply helpers/mathjax.js:76]: That's right, yeah.

File: components/html-block/test/html-block.test.js
Comment (grant-cleary) [reply components/html-block/test/html-block.test.js:164]: I certainly think there's value here. Unit test might be harder because I'm not quite sure how easy it is for us to actually await the rendering before asserting anything. The html-block doesn't currently broadcast when it's done its work. Visual diff is easier since we're inevitably tossing waits around anyways.

**Comment by grant-cleary** [comment]
Going to introduce the visual diff test in a separate commit.

File: helpers/mathjax.js
Comment (dbatiste) [reply helpers/mathjax.js]: If it's invalid otherwise then it's all good. If there is a chance to unexpected clobber, then we should play it safe.

Comment (grant-cleary) [reply helpers/mathjax.js]: Yeah, it shouldn't be valid for there to be styling on there because MathJax will just wipe it out anyways, but it's a trivial change and strictly safer, so... 🤷

File: helpers/mathjax.js
Comment (grant-cleary) [reply helpers/mathjax.js:77]: Technically I suppose that makes the return value completely pointless, since the reference to it in the html-block rendering code will just implicitly update.

Comment (dbatiste) [reply helpers/mathjax.js:77]: Yeah, and also a point of failure as we've been chatting about.  If all the renderers currently return the element it was passed, maybe `d2l-html-block` shouldn't expect it to be returned, and instead assume that all processing will be done on the element it passes. Not sure if this would be limiting in the future, as the previous version of MathJax renderer required the ability to return a new element. 🤔

File: components/html-block/test/html-block.test.js
Comment (dbatiste) [reply components/html-block/test/html-block.test.js:164]: If visual-diff covers it, I wouldn't create a unit test.

File: helpers/mathjax.js
Comment (grant-cleary) [reply helpers/mathjax.js:77]: Yeah, that was my main thought. Certainly right now the existing renderers act on the elements as they are, and enforcing that in the component makes it very obvious that we're expecting renderers to run inline. I don't think it's particularly difficult to reverse course later if we want. Realistically we would just switch the renderers to start returning the element and then adjust the html-block to start consuming it again. I'm certainly fine removing the return value for now though.

Comment (grant-cleary) [reply helpers/mathjax.js]: Ok, I made this change in the core renderers and the html-block component itself. It will need to be reflected in the @mentions renderer as well, but because that's in a different repo that change will have to wait until post-merge.

Comment (dbatiste) [reply helpers/mathjax.js]: Cool.  Thanks for updating that.  It would be quite unfortunate if a plugin renderer forgot to return the element and in effect broken the component or other renderers.

Comment (grant-cleary) [reply helpers/mathjax.js]: Haha yeah, I was trying to think if any renderer could have a valid use case for wanting to return something, but realistically if a renderer needs to detach the element from the DOM or something, it's also going to need to add it back as well. At least this way the requirement is super clear.

File: helpers/mathjax.js
Comment (grant-cleary) [inline helpers/mathjax.js:70]: I was curious to see if this works and it seems to now. I think the trick is the promise. MathJax grabs the root node of whatever elements it's working on and stores that information globally. I think if multiple renders collided, you could get into a situation where that document reference could change mid-render and MathJax would start typesetting things it shouldn't have typeset.

File: helpers/mathjax.js
Comment (grant-cleary) [inline helpers/mathjax.js:81]: @dbatiste I was musing about Lit's `adoptStyles` here when you sent me that DM about it. In theory we could use that here to attach those styles directly to the `shadowRoot` if we so wished.

**Comment by dbatiste** [comment]
This is looking pretty good.  Is there more MathJax stuff that you are hoping to change?

**Comment by grant-cleary** [comment]
> This is looking pretty good. Is there more MathJax stuff that you are hoping to change?

I'm taking a brief look at whether it's possible to get rid of the `mjx-doc` structure (though I'm not optimistic). Beyond that, I think this will probably be good to go.

**Comment by grant-cleary** [comment]
> I'm taking a brief look at whether it's possible to get rid of the `mjx-doc` structure (though I'm not optimistic). Beyond that, I think this will probably be good to go.

Yeah, this was a wash. As far as I'm concerned this PR is likely in its end state.

**Comment by grant-cleary** [comment]
Deliberately not going to merge this until 20.22.10 branching so the MathJax version bump has a month to soak first.

**Review by dbatiste — APPROVED** [review]

**Pull request merged**
